libp2p/src/builder.rs

```diff
@@ -0,0 +1,1203 @@
+}
+
+pub struct QuicBuilder<P, T> {
+    transport: T,
+    keypair: libp2p_identity::Keypair,
+    phantom: PhantomData<P>,
+}
+
+#[cfg(all(feature = "quic", feature = "async-std"))]
+impl<T: AuthenticatedMultiplexedTransport> QuicBuilder<AsyncStd, T> {
+    pub fn with_quic(self) -> RelayBuilder<AsyncStd, impl AuthenticatedMultiplexedTransport> {
+        RelayBuilder {
+            transport: self
+                .transport
+                .or_transport(
+                    libp2p_quic::async_std::Transport::new(libp2p_quic::Config::new(&self.keypair))
+                        .map(|(peer_id, muxer), _| (peer_id, StreamMuxerBox::new(muxer))),
+                )
+                .map(|either, _| either.into_inner()),
+            keypair: self.keypair,
+            phantom: PhantomData,
+        }
+    }
+}
+
+#[cfg(all(feature = "quic", feature = "tokio"))]
+impl<T: AuthenticatedMultiplexedTransport> QuicBuilder<Tokio, T> {
+    pub fn with_quic(self) -> RelayBuilder<Tokio, impl AuthenticatedMultiplexedTransport> {
+        RelayBuilder {
+            transport: self
+                .transport
+                .or_transport(
+                    libp2p_quic::tokio::Transport::new(libp2p_quic::Config::new(&self.keypair))
+                        .map(|(peer_id, muxer), _| (peer_id, StreamMuxerBox::new(muxer))),
+                )
+                .map(|either, _| either.into_inner()),
+            keypair: self.keypair,
+            phantom: PhantomData,
+        }
+    }
+}
+
+impl<P, T> QuicBuilder<P, T> {
+    pub fn without_quic(self) -> RelayBuilder<P, T> {
+        RelayBuilder {
+            transport: self.transport,
+            keypair: self.keypair,
+            phantom: PhantomData,
+        }
+    }
+}
+
+// Shortcuts
+impl<P, T: AuthenticatedMultiplexedTransport> QuicBuilder<P, T> {
+    #[cfg(feature = "relay")]
+    pub fn with_relay(self) -> RelayTlsBuilder<P, T> {
+        RelayTlsBuilder {
+            transport: self.transport,
+            keypair: self.keypair,
+            phantom: PhantomData,
+        }
+    }
+
+    pub fn with_other_transport<OtherTransport: AuthenticatedMultiplexedTransport>(
+        self,
+        constructor: impl FnMut(&libp2p_identity::Keypair) -> OtherTransport,
+    ) -> OtherTransportBuilder<P, impl AuthenticatedMultiplexedTransport, NoRelayBehaviour> {
+        self.without_quic()
+            .without_relay()
+            .with_other_transport(constructor)
+    }
+
+    #[cfg(feature = "websocket")]
+    pub fn with_websocket(
+        self,
+    ) -> WebsocketTlsBuilder<P, impl AuthenticatedMultiplexedTransport, NoRelayBehaviour> {
+        self.without_quic()
+            .without_relay()
+            .without_any_other_transports()
+            .without_dns()
+            .with_websocket()
+    }
+
+    pub fn with_behaviour<B>(
+        self,
+        constructor: impl FnMut(&libp2p_identity::Keypair) -> Result<B, Box<dyn std::error::Error>>,
```

Comment:
Can we avoid the `Ok` wrapping if we do something like:


```suggestion
        constructor: impl FnMut(&libp2p_identity::Keypair) -> O,
```
and

```rust
where O: Into<Result<B, E>>
```

And then provide an implementation like: 

```rust
impl Into<Result<B, Infallible>> for B where B: NetworkBehaviour { }
```